torchao/dtypes/nf4tensor.py

```diff
@@ -37,11 +40,234 @@ def decorator(func):
     return decorator
 
 
-@implements([torch.ops.aten.detach.default, torch.ops.aten.detach])
+@implements([torch.ops.aten.detach])
 def noop_detach(func, *args, **kwargs):
     return args[0][0]
 
 
+@implements(
+    [
+        aten.detach.default,
+    ]
+)
+def nf4_detach(aten_op, args, kwargs=None):
```

Comment:
Dumb q : whats the difference between this detach and the previous, ?

Reply:
I might just didnt get the point yet but allow me write down the silly explanation:  previous one (if you mean what's on the left) is noop since it's returning directly. the newly implemented one is applying detach to inner tensors.

Reply:
So essentially the previous detach was incorrect?

I think I wrote it as a no-op since NF4Tensor shoudlnt every be produced from tensors that requries grad and thus doesnt need to be detached from the graph. Cant remmber though if this is an aliasing op

Reply:
>  NF4Tensor shoudlnt every be produced from tensors

noob: is it safe to claim NF4 quantization is for non-trainable parameters? thus it's always on freezed parameter in finetuning or inference? 

If yes, I guess wehther detach is always equivalent to no-op because `requires_grad=False`

Reply:
If we make that assumption that `requires_grad=False` and `detach` is a no-op, can we add an assertion that checks for `args[0].requires_grad`?

Also, I am not sure that we need to detach all inner tensors. cc: @bdhirsh

Reply:
> we add an assertion that checks for args[0].requires_grad

there is `# assert not inpt_tensor.requires_grad` in `from_tensor` from @drisspg. Want to check if it's commented out for some reason

Reply:
Kind of long and pedantic answer below:

I think if you want to be completely correct: `y = x.detach()` means that `y` and `x` are distinct tensors that alias each other (`y` also has its autograd metadata scrubbed but that part won't matter if NF4 isn't used in training). Distinct tensors mean that metadata mutations are not shared between the tensors (if you call `y.t_()`, that will not transpose the dims of `x`).

Technically none of this matters if you know that there will be no mutations in your program, but if you want to make sure that you don't have any silent correctness issues if your subclass encounters mutations, then you'll need to:

(1) call `y_inner = x_inner.detach()` on the inner tensor to preserve the aliasing behavior on the inner tensors.
(2) alias the storages of the NF4 output with the input. Today there's an API that you can call to handle this, `return_and_correct_aliasing`: https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/two_tensor.py#L74

P.S. I think there's some work we could probably do to improve this situation, since aliasing is complicated (for example if we ran functionalization **above** your subclass, then you wouldn't have to worry about the aliasing of your inner tensors being correct for view ops).

Reply:
thanks for explaining in detail. will take a look at ‘return_and_correct_aliasing’. new concept for me